AK/DistinctNumeric.h

```diff
@@ -0,0 +1,314 @@
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <AK/Types.h>
+
+namespace AK {
+
+namespace internal {
+template<int N>
+constexpr int literal_length(const char (&)[N])
+{
+    return N - 1;
+}
+}
+
+/**
+ * This implements a "distinct" numeric type that is intentionally incompatible
+ * to other incantations. The intention is that each "distinct" type that you
+ * want simply gets different values for `fn_length` and `line`. The macros
+ * `TYPEDEF_DISTINCT_NUMERIC_*()` at the bottom of `DistinctNumeric.h`.
+ *
+ * `seq`, `cmp`, `truthy`, `flags`, `shift`, and `arith` simply split up the
+ * space of operators into 6 simple categories:
+ * - No matter the values of these, `DistinctNumeric` always implements `==` and `!=`.
+ * - If `seq` is true, then `++a`, `a++`, `--a`, and `a--` are implemented.
+ * - If `cmp` is true, then `a>b`, `a<b`, `a>=b`, and `a<=b` are implemented.
+ * - If `truthy` is true, then `!a`, `a&&b`, and `a||b` are implemented (through `operator bool()`.
+ * - If `flags` is true, then `~a`, `a&b`, `a|b`, `a^b`, `a&=b`, `a|=b`, and `a^=b` are implemented.
+ * - If `shift` is true, then `a<<b`, `a>>b`, `a<<=b`, `a>>=b` are implemented.
+ * - If `arith` is true, then `a+b`, `a-b`, `+a`, `-a`, `a*b`, `a/b`, `a%b`, and the respective `a_=b` versions are implemented.
+ * The semantics are always those of the underlying basic type `T`.
+ *
+ * These can be combined arbitrarily. Want a numeric type that supports `++a`
+ * and `a >> b` but not `a > b`? Sure thing, just set
+ * `seq=true, cmp=false, shift=true` and you're done!
+ * Furthermore, some of these overloads make more sense with specific types, like `a&&b` which should be able to operate
+ *
+ * I intentionally decided against overloading `&a` because these shall remain
+ * numeric types.
+ *
+ * The C++20 `operator<=>` would require, among other things `std::weak_equality`.
+ * Since we do not have that, it cannot be implemented.
+ *
+ * The are many operators that do not work on `int`, so I left them out:
+ * `a[b]`, `*a`, `a->b`, `a.b`, `a->*b`, `a.*b`.
```

Comment:
Not sure if any of these are even useful to have for _numeric_ types, but you can probably use a little template magic (or `requires`, we have concepts...kinda) like `auto helper(a, b) -> decltype(a.b)`...probably.

Reply:
Interesting! But I intend to stick to "just" numeric types. However, I expect to run into something like `some_vector[my_distinct_numeric]`, so I'll probably have to think hard about that, eventually.